File: generate_data.py
import csv
import random
from datetime import datetime, timedelta
import os

random.seed(42)

NUM_ROWS = 50000
FRAUD_RATIO = 0.015
NUM_FRAUD = int(NUM_ROWS * FRAUD_RATIO)
NUM_NORMAL = NUM_ROWS - NUM_FRAUD
NUM_USERS = 5000

user_ids = [f"U{str(i).zfill(5)}" for i in range(1, NUM_USERS + 1)]

normal_merchants = ['Groceries', 'Dining', 'Retail', 'Transportation', 'Entertainment', 'Subscription', 'Utilities']
fraud_merchants = ['Electronics', 'Jewelry', 'Crypto', 'Luxury Goods', 'Transfer']

def generate_normal_amount():
    amt = random.expovariate(1/50) + 10 # mean ~60
    return round(min(amt, 900), 2)

def generate_transaction_data():
    data = []
    
    # Generate Normal Transactions
    for _ in range(NUM_NORMAL):
        user = random.choice(user_ids)
        amount = generate_normal_amount()
        
        days_ago = random.randint(0, 30)
        hour = random.choice(range(6, 24))
        minute = random.randint(0, 59)
        timestamp = datetime.now() - timedelta(days=days_ago, hours=24-hour, minutes=minute)
        
        merchant = random.choice(normal_merchants)
        location = random.choice(['New York, NY', 'Los Angeles, CA', 'Chicago, IL', 'Houston, TX', 'Phoenix, AZ'])
        device = f"DEV_{user}_1"
        
        data.append({'user_id': user, 'amount': amount, 'time': timestamp, 'merchant': merchant, 'location': location, 'device': device, 'is_fraud': 0})
        
    # Generate Fraud Transactions
    for _ in range(NUM_FRAUD):
        user = random.choice(user_ids)
        
        # Fraud pattern 1: Unusually large transactions
        amount = round(random.uniform(1000, 5000), 2)
        
        # Fraud pattern 2: Unusual times (e.g. 1 AM - 5 AM)
        days_ago = random.randint(0, 30)
        hour = random.randint(0, 5)
        minute = random.randint(0, 59)
        timestamp = datetime.now() - timedelta(days=days_ago, hours=24-hour, minutes=minute)
        
        # Fraud pattern 3: Rare/high-value merchants
        merchant = random.choice(fraud_merchants)
        
        # Fraud pattern 4: Sudden location change
        location = random.choice(['Miami, FL', 'Seattle, WA', 'International_Unknown'])
        
        # Fraud pattern 5: New device usage
        device = f"DEV_UNKNOWN_{random.randint(1000, 9999)}"
        
        data.append({'user_id': user, 'amount': amount, 'time': timestamp, 'merchant': merchant, 'location': location, 'device': device, 'is_fraud': 1})
        
    # Shuffle data
    random.shuffle(data)
    
    # Assign Tx ID and prepare rows
    rows = []
    for i, row in enumerate(data, 1):
        tx_id = f"TXN{str(i).zfill(6)}"
        rows.append([
            tx_id, 
            row['user_id'], 
            row['amount'], 
            row['time'].strftime("%Y-%m-%d %H:%M:%S"), 
            row['merchant'], 
            row['location'], 
            row['device'], 
            row['is_fraud']
        ])
        
    return rows

if __name__ == "__main__":
    print("Generating dataset without external dependencies...")
    rows = generate_transaction_data()
    
    output_dir = r"c:\USM\VHACK_2026\VHack_2026\fraud-detection-system\backend\data"
    os.makedirs(output_dir, exist_ok=True)
    
    output_file = os.path.join(output_dir, "transactions.csv")
    with open(output_file, 'w', newline='', encoding='utf-8') as f:
        writer = csv.writer(f)
        writer.writerow(['transaction_id', 'user_id', 'transaction_amount', 'transaction_time', 'merchant_category', 'location', 'device_id', 'is_fraud'])
        writer.writerows(rows)
        
    fraud_count = sum(1 for r in rows if r[8] == 1)
    
    print(f"Dataset saved to {output_file}")
    print(f"Total rows: {len(rows)}")
    print(f"Fraud ratio: {(fraud_count / len(rows)) * 100:.2f}%")

File: generate_data.js
const fs = require('fs');
const path = require('path');

const NUM_ROWS = 50000;
const FRAUD_RATIO = 0.015;
const NUM_FRAUD = Math.floor(NUM_ROWS * FRAUD_RATIO);
const NUM_NORMAL = NUM_ROWS - NUM_FRAUD;
const NUM_USERS = 5000;

// simple seedable prng
let seed = 42;
function random() {
    let t = seed += 0x6D2B79F5;
    t = Math.imul(t ^ t >>> 15, t | 1);
    t ^= t + Math.imul(t ^ t >>> 7, t | 61);
    return ((t ^ t >>> 14) >>> 0) / 4294967296;
}

const user_ids = Array.from({length: NUM_USERS}, (_, i) => `U${String(i+1).padStart(5, '0')}`);

const normal_merchants = ['Groceries', 'Dining', 'Retail', 'Transportation', 'Entertainment', 'Subscription', 'Utilities'];
const fraud_merchants = ['Electronics', 'Jewelry', 'Crypto', 'Luxury Goods', 'Transfer'];
const normal_locations = ['New York, NY', 'Los Angeles, CA', 'Chicago, IL', 'Houston, TX', 'Phoenix, AZ'];
const fraud_locations = ['Miami, FL', 'Seattle, WA', 'International_Unknown'];

function generate_transaction_data() {
    const data = [];
    const now = new Date();
    
    // Normal transactions
    for (let i = 0; i < NUM_NORMAL; i++) {
        const user = user_ids[Math.floor(random() * user_ids.length)];
        
        // Exponetial-like random amount
        const amount = Math.min(Math.round((-Math.log(1 - random()) * 50 + 10) * 100) / 100, 900);
        
        const daysAgo = Math.floor(random() * 30);
        const hour = Math.floor(random() * 18) + 6; // 6 to 23
        const minute = Math.floor(random() * 60);
        const timestamp = new Date(now.getTime() - (daysAgo * 86400000) - ((24 - hour) * 3600000) - (minute * 60000));
        
        const merchant = normal_merchants[Math.floor(random() * normal_merchants.length)];
        const location = normal_locations[Math.floor(random() * normal_locations.length)];
        const device = `DEV_${user}_1`;
        
        data.push({
            user_id: user, amount, time: timestamp, merchant, location, device, is_fraud: 0
        });
    }
    
    // Fraud transactions
    for (let i = 0; i < NUM_FRAUD; i++) {
        const user = user_ids[Math.floor(random() * user_ids.length)];
        
        const amount = Math.round((random() * 4000 + 1000) * 100) / 100; // 1000 to 5000
        
        const daysAgo = Math.floor(random() * 30);
        const hour = Math.floor(random() * 6); // 0 to 5
        const minute = Math.floor(random() * 60);
        const timestamp = new Date(now.getTime() - (daysAgo * 86400000) - ((24 - hour) * 3600000) - (minute * 60000));
        
        const merchant = fraud_merchants[Math.floor(random() * fraud_merchants.length)];
        const location = fraud_locations[Math.floor(random() * fraud_locations.length)];
        const device = `DEV_UNKNOWN_${Math.floor(random() * 9000) + 1000}`;
        
        data.push({
            user_id: user, amount, time: timestamp, merchant, location, device, is_fraud: 1
        });
    }
    
    // Shuffle using Fisher-Yates
    for (let i = data.length - 1; i > 0; i--) {
        const j = Math.floor(random() * (i + 1));
        [data[i], data[j]] = [data[j], data[i]];
    }
    
    // Format rows
    const rows = data.map((row, i) => {
        const tx_id = `TXN${String(i+1).padStart(6, '0')}`;
        const timeStr = row.time.toISOString().replace('T', ' ').substring(0, 19);
        return `${tx_id},${row.user_id},${row.amount},${timeStr},${row.merchant},"${row.location}",${row.device},${row.is_fraud}`;
    });
    
    return rows;
}

const rows = generate_transaction_data();
const outputDir = path.join('C:', 'USM', 'VHACK_2026', 'VHack_2026', 'fraud-detection-system', 'backend', 'data');
fs.mkdirSync(outputDir, { recursive: true });

const header = 'transaction_id,user_id,transaction_amount,transaction_time,merchant_category,location,device_id,is_fraud';
fs.writeFileSync(path.join(outputDir, 'transactions.csv'), header + '\n' + rows.join('\n'));

console.log(`Generated ${rows.length} rows.`);
